packages/cdk8s-plus-17/src/statefulset.ts

```diff
@@ -0,0 +1,217 @@
+ * rescheduling.
+ *
+ * If you want to use storage volumes to provide persistence for your workload, you
+ * can use a StatefulSet as part of the solution. Although individual Pods in a StatefulSet
+ * are susceptible to failure, the persistent Pod identifiers make it easier to match existing
+ * volumes to the new Pods that replace any that have failed.
+ *
+ * Using StatefulSets
+ * ------------------
+ * StatefulSets are valuable for applications that require one or more of the following.
+ *
+ * - Stable, unique network identifiers.
+ * - Stable, persistent storage.
+ * - Ordered, graceful deployment and scaling.
+ * - Ordered, automated rolling updates.
+ */
+export class StatefulSet extends Resource implements IPodTemplate {
+  /**
+    * Number of desired pods.
+    */
+  public readonly replicas: number;
+
+  /**
+    * Management policy to use for the set.
+    */
+  public readonly podManagementPolicy: PodManagementPolicy;
+
+  /**
+    * @see base.Resource.apiObject
+    */
+  protected readonly apiObject: ApiObject;
+
+  private readonly _podTemplate: PodTemplate;
+  private readonly _labelSelector: Record<string, string>;
+
+  private readonly _service: Service;
+
+  constructor(scope: Construct, id: string, props: StatefulSetProps = {}) {
+    super(scope, id, { metadata: props.metadata });
+
+    this.apiObject = new k8s.KubeStatefulSet(this, 'Resource', {
+      metadata: props.metadata,
+      spec: Lazy.any({ produce: () => this._toKube() }),
+    });
+
+    if (!props.servicePort && !props.service?.ports) {
+      throw new Error('Cannot build a stateful set with a Service without ports.');
+    }
+
+    this._service = new Service(this, 'Service', {
+      metadata: props.serviceName ? { name: props.serviceName } : undefined,
+      type: props.service?.type ?? ServiceType.CLUSTER_IP,
+      clusterIP: props.service?.clusterIP,
+      ports:
+            props.service?.ports ?? (props.servicePort != null ? [{ port: props.servicePort }] : []),
+    });
+    this.apiObject.addDependency(this._service);
+
+    this.replicas = props.replicas ?? 1;
+    this.podManagementPolicy = props.podManagementPolicy ?? PodManagementPolicy.ORDERED_READY;
+    this._podTemplate = new PodTemplate(props);
+    this._labelSelector = {};
+
+    if (props.defaultSelector ?? true) {
+      const selector = 'cdk8s.statefulset';
+      const matcher = Names.toLabelValue(this);
+      this.podMetadata.addLabel(selector, matcher);
+      this.selectByLabel(selector, matcher);
+    }
+
+    const selectors = Object.entries(this.labelSelector);
+    for (const [k, v] of selectors) {
+      this._service.addSelector(k, v);
+    }
+  }
+
+  get podMetadata(): ApiObjectMetadataDefinition {
+    return this._podTemplate.podMetadata;
+  }
+
+  /**
+    * The labels this deployment will match against in order to select pods.
+    *
+    * Returns a a copy. Use `selectByLabel()` to add labels.
+    */
+  get labelSelector(): Record<string, string> {
```

Comment:
To be consistent:

```suggestion
public  get labelSelector(): Record<string, string> {
```

Reply:
Got it.  I am so used to my production projects just adding this type of thing automatically using prettier and tslint that I forgot.  Sorry.